### PR TITLE
HPCC-16961 Automatically provide sample JSON responses for ESP services

### PR DESCRIPTION
**Pull request by afishbeck**
Generated forms will now have a link to Sample JSON response messages.

Navigating to the service page, for example: http://IP:8010/WsWorkunits
now has links that will provide a collection of all the sample messages
for all the methods in the given service.

Signed-off-by: Anthony Fishbeck <anthony.fishbeck@lexisnexis.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [ ] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [x] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by richardkchapman** [comment]
@wangkx Please review

File: esp/bindings/http/platform/httpbinding.cpp [hunk 1]
```diff
@@ -1673,21 +1778,24 @@ int EspHttpBinding::onGetIndex(IEspContext &context, CHttpRequest* request,  CHt
...
+        for(int i = 0, tot = methods.length(); i < tot; ++i)
+        {
+            CMethodInfo &info = methods.item(i);
             {
-                CMethodInfo &info = methods.item(i);
-                {
-                    wsLink.clear().appendf("%s%s", methods.item(i).m_label.str(),urlFormParams.str());
-                    list->appendContent(new CHtmlLink(methods.item(i).m_label.str(), wsLink.str()));
-                }
+                wsLink.clear().appendf("%s%s", methods.item(i).m_label.str(),urlFormParams.str());
```
Comment (wangkx) [inline esp/bindings/http/platform/httpbinding.cpp]: Use setf()?

File: esp/bindings/http/platform/httpbinding.cpp [hunk 2]
```diff
@@ -1673,21 +1778,24 @@ int EspHttpBinding::onGetIndex(IEspContext &context, CHttpRequest* request,  CHt
...
-            StringBuffer urlFormParams(urlParams);
-            urlFormParams.append(urlFormParams.length()>0 ? "&form" : "?form");
-            for(int i = 0, tot = methods.length(); i < tot; ++i)
+        //links to the form pages
+        list = (CHtmlList *)page.appendContent(new CHtmlList);
+        StringBuffer urlFormParams(urlParams);
+        urlFormParams.append(urlFormParams.length()>0 ? "&form" : "?form");
+        for(int i = 0, tot = methods.length(); i < tot; ++i)
+        {
+            CMethodInfo &info = methods.item(i);
```
Comment (wangkx) [inline esp/bindings/http/platform/httpbinding.cpp]: The info not used?

File: esp/services/common/jsonhelpers.hpp [hunk 3]
```diff
@@ -266,24 +286,46 @@ namespace JsonHelpers
...
                     const char *childname = type->queryFieldName(idx);
-                    if (reqTree)
-                        childtree = reqTree->queryPropTree(childname);
-                    buildJsonMsg(parentTypes, type->queryFieldType(idx), out, childname, childtree, flags & ~REQSF_ROOT);
+                    bool repeats = type->queryFieldRepeats(idx);
+                    if (repeats)
+                    {
+                        IPropertyTreeIterator *children = NULL;
+                        if (reqTree)
+                            children = reqTree->getElements(childname);
```
Comment (wangkx) [inline esp/services/common/jsonhelpers.hpp]: setown?

**Review by wangkx — COMMENTED** [review]
@afishbeck back to you.

**Comment by afishbeck** [comment]
@wangkx good catches, I've updated please check.  I'll squash after you approve.

**Comment by wangkx** [comment]
@afishbeck looks fine.

**Comment by richardkchapman** [comment]
@afishbeck Please squash

**Comment by afishbeck** [comment]
@richardkchapman squashed.

**Pull request merged**